kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ActionsIterator.java

```diff
@@ -110,7 +117,18 @@ public ActionWrapper next() {
             throw new NoSuchElementException("No next element");
         }
 
-        return actionsIter.get().next();
+        ActionWrapper next = actionsIter.get().next();
+
+        // If the ActionWrapper schema does not match the read schema, then we have injected sidecar
+        // files into the schema. Append any sidecar files found to the end of the file list and
+        // return a new ActionWrapper with the sidecar files removed from the read schema.
+        if (next.getContainsSidecarsInSchema() && next.isFromCheckpoint()) {
```

Comment:
Do we expect `getContainsSidecarsInSchema` to be true and `isFromCheckpoint` to be false? Should we add an assertion somewhere to catch such scenario early?

Reply:
This should not be possible (SidecarFile actions are only injected into the schema for V2/classic top-level checkpoints that might contain these actions and when we are looking for AddFile/RemoveFile actions). Changed from condition to assertion.